wallet-new/src/Cardano/Wallet/Kernel/Addresses.hs

```diff
@@ -0,0 +1,158 @@
+module Cardano.Wallet.Kernel.Addresses (
+    createAddress
+    -- * Errors
+    , CreateAddressError(..)
+    ) where
+
+import           Universum
+
+import           Control.Lens (to)
+import           Formatting (bprint, (%))
+import qualified Formatting as F
+import qualified Formatting.Buildable
+import           System.Random.MWC (GenIO, createSystemRandom, uniformR)
+
+import           Data.Acid (update)
+
+import           Pos.Core (Address, IsBootstrapEraAddr (..), deriveLvl2KeyPair)
+import           Pos.Crypto (EncryptedSecretKey, PassPhrase,
+                     ShouldCheckPassphrase (..))
+
+import           Cardano.Wallet.Kernel (PassiveWallet, walletKeystore, wallets)
+import           Cardano.Wallet.Kernel.DB.AcidState (CreateHdAddress (..))
+import           Cardano.Wallet.Kernel.DB.HdWallet (HdAccountId,
+                     HdAccountIx (..), HdAddressId (..), HdAddressIx (..),
+                     UnknownHdAccount (..), hdAccountIdIx, hdAccountIdParent,
+                     hdAddressIdIx)
+import           Cardano.Wallet.Kernel.DB.HdWallet.Create
+                     (CreateHdAddressError (..), initHdAddress)
+import           Cardano.Wallet.Kernel.DB.HdWallet.Derivation
+                     (HardeningMode (..), deriveIndex)
+import           Cardano.Wallet.Kernel.DB.InDb (InDb (..))
+import qualified Cardano.Wallet.Kernel.Keystore as Keystore
+import           Cardano.Wallet.Kernel.Types (AccountId (..), WalletId (..))
+
+import           Test.QuickCheck (Arbitrary (..), oneof)
+
+data CreateAddressError =
+      CreateAddressUnknownHdAccount UnknownHdAccount
+      -- ^ When trying to create the 'Address', the parent 'Account' was not
+      -- there.
+    | CreateAddressKeystoreNotFound AccountId
+      -- ^ When trying to create the 'Address', the 'Keystore' didn't have
+      -- any secret associated with this 'Account'.
+      -- there.
+    | CreateAddressHdCreationFailed CreateHdAddressError
+      -- ^ The creation of the new HD 'Address' failed with a database error.
+    | CreateAddressHdRndGenerationFailed HdAccountId
+      -- ^ The crypto-related part of address generation failed.
+    | CreateAddressHdRndAddressSpaceSaturated HdAccountId
+      -- ^ The available number of HD addresses in use in such that trying
+      -- to find another random index would be too expensive
+    deriving Eq
+
+-- TODO(adn)
+instance Arbitrary CreateAddressError where
```

Comment:
I have added most of the required instances in my pending work on `newTransaction`, I left the `oneof []` here to fill later as part of that work.